### ISSUE TITLE
Can't the lock be removed ?

### ISSUE THREAD

**Issue by bilal-fazlani** [bilal-fazlani/tracker-enabled-dbcontext#13]
```
private static TValue GetFromCache<TKey, TValue>(ConcurrentDictionary<TKey, TValue> dictionary, TKey key,
            Func<TValue> func)
        {
            lock (dictionary)
            {
                if (!dictionary.ContainsKey(key))
                {
                    dictionary[key] = func();
                }

                return dictionary[key];
            }
        }
```

The dictionary is `ConcurrentDictionary`. Can't the lock be removed ? What could be side-effect of removing lock on a `ConcurrentDictionary` ? I thought they are thread-safe.

**Comment by GeertvanHorrik** [comment]
I think it cannot be removed. Just because the dictionary is thread-safe, the check and the return should be a single atomic action, not 3 different ones (the values might change in-between by different threads. We must keep the lock because even a concurrent dictionary is unaware of the calling code and the logic. It only means the values can safely be accessed by different threads.

**Comment by bilal-fazlani** [comment]
Ok. cool. Another thing I noticed was, the lambda expression (anonymous method) being passed to 'GetFromCache()' gets executed even before 'func();' is called. And this happens even if key is present in dictionary. which was killing the entire purpose of caching. 
After I converted the lambda in to a private method and passed it, it worked as expected. 
I must say though, just by looking at your code, I learnt a lot :)

**Comment by GeertvanHorrik** [comment]
Strange. A lambda is the same as a function. Did you try it with a Console.WriteLine? Never trust the debugger, especially on lambdas. You need to set the breakpoint very carefully (set the cursor inside the lambda, then hit f9).

**Comment by bilal-fazlani** [comment]
I did set the breakpoint inside lambda and it go hit before GetFromCache(). But yes, it's worth testing with console.writeline(). Will do that and let you know.

**Comment by bilal-fazlani** [comment]
@GeertvanHorrik 
Strange indeed, When tried in a console POC application, the behavior of lambda was as expected. 

Anyways, About removing of the locks, I did some research, I found that ConcurrentCollection does expose threadsafe methods, which will allow us to remove the locks. The one we can use is [GetOrAdd](https://msdn.microsoft.com/en-us/library/ee378677(v=vs.110).aspx). I am guessing this method locks the dictionary internally.
